// keep-a-changelog/script/bin.js
#!/usr/bin/env node
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
require("./_dnt.polyfills.js");
const dntShim = __importStar(require("./_dnt.shims.js"));
const mod_js_1 = require("./deps/jsr.io/@std/path/1.0.8/mod.js");
const mod_js_2 = require("./mod.js");
const parse_args_js_1 = require("./deps/jsr.io/@std/cli/1.0.13/parse_args.js");
const mod_js_3 = require("./deps/jsr.io/@std/ini/0.225.2/mod.js");
const settings_js_1 = __importDefault(require("./src/settings.js"));
const argv = (0, parse_args_js_1.parseArgs)(dntShim.Deno.args, {
    default: {
        file: "CHANGELOG.md",
        format: "compact",
        release: null,
        create: null,
        url: null,
        https: true,
        quiet: false,
        head: null,
        combine: false,
        "bullet-style": "-",
    },
    string: ["file", "format", "url", "head", "bullet-style"],
    boolean: [
        "https",
        "init",
        "latest-release",
        "latest-release-full",
        "quiet",
        "help",
        "combine",
        "no-v-prefix",
        "no-sort-releases",
    ],
    alias: {
        h: "help",
    },
});
const file = (0, mod_js_1.join)(dntShim.Deno.cwd(), argv.file);
try {
    if (argv.help) {
        showHelp();
        dntShim.Deno.exit(0);
    }
    if (argv.init) {
        const changelog = new mod_js_2.Changelog("Changelog").addRelease(new mod_js_2.Release("0.1.0", new Date(), "First version"));
        changelog.format = argv.format;
        changelog.bulletStyle = argv["bullet-style"];
        save(file, changelog, true);
        dntShim.Deno.exit(0);
    }
    const changelog = (0, mod_js_2.parser)(dntShim.Deno.readTextFileSync(file), {
        autoSortReleases: !argv["no-sort-releases"],
    });
    changelog.format = argv.format;
    changelog.bulletStyle = argv["bullet-style"];
    if (argv["no-v-prefix"]) {
        changelog.tagNameBuilder = (release) => String(release.version);
    }
    if (argv["latest-release"]) {
        const release = changelog.releases.find((release) => release.date && release.version);
        if (release) {
            console.log(release.version?.toString());
        }
        dntShim.Deno.exit(0);
    }
    if (argv["latest-release-full"]) {
        const release = changelog.releases.find((release) => release.date && release.version);
        if (release) {
            console.log(release.toString());
        }
        dntShim.Deno.exit(0);
    }
    if (argv.release) {
        const release = changelog.releases.find((release) => {
            if (release.date) {
                return false;
            }
            if (typeof argv.release === "string") {
                return !release.version || argv.release === release.version.toString();
            }
            return !!release.version;
        });
        if (release) {
            release.date = new Date();
            if (typeof argv.release === "string") {
                release.setVersion(argv.release);
            }
        }
        else {
            console.error("Not found any valid unreleased version");
            dntShim.Deno.exit(1);
        }
    }
    if (argv.combine) {
        const combinedReleases = changelog.releases.reduce((acc, release) => {
            if (release.version) {
                if (acc[release.version]) {
                    acc[release.version].combineChanges(release.changes);
                }
                else {
                    acc[release.version] = release;
                }
            }
            return acc;
        }, {});
        changelog.releases = Object.values(combinedReleases);
    }
    if (argv.create) {
        let version = typeof argv.create === "string" ? argv.create : undefined;
        if (version === "major" || version === "minor" || version === "patch") {
            const latestRelease = changelog.releases.find((release) => release.parsedVersion);
            if (!latestRelease) {
                console.error("No releases found to bump version from.");
                dntShim.Deno.exit(1);
            }
            let { major, minor, patch } = latestRelease.parsedVersion;
            if (version === "major") {
                major += 1;
                minor = 0;
                patch = 0;
            }
            else if (version === "minor") {
                minor += 1;
                patch = 0;
            }
            else if (version === "patch") {
                patch += 1;
            }
            version = `${major}.${minor}.${patch}`;
        }
        const release = changelog.releases.find((release) => release.version === version);
        if (release) {
            console.warn("Release already exists.");
        }
        else {
            changelog.addRelease(new mod_js_2.Release(version));
        }
    }
    save(file, changelog);
}
catch (err) {
    console.error(red(err.message));
    if (!argv.quiet) {
        dntShim.Deno.exit(1);
    }
}
function save(file, changelog, isNew = false) {
    changelog.url = argv.url || changelog.url || getRemoteUrl(argv.https);
    if (!changelog.url) {
        console.error(red('Please, set the repository url with --url="https://github.com/username/repository"'));
        changelog.url = "https://example.com";
    }
    if (changelog.url) {
        const settings = (0, settings_js_1.default)(changelog.url);
        if (settings) {
            changelog.head = settings.head;
            changelog.tagLinkBuilder = settings.tagLink;
        }
    }
    if (argv.head) {
        changelog.head = argv.head;
    }
    dntShim.Deno.writeTextFileSync(file, changelog.toString());
    if (isNew) {
        console.log(green("Generated new file"), file);
    }
    else {
        console.log(green("Updated file"), file);
    }
}
function red(message) {
    return "\u001b[" + 31 + "m" + message + "\u001b[" + 39 + "m";
}
function green(message) {
    return "\u001b[" + 32 + "m" + message + "\u001b[" + 39 + "m";
}
function normalizeUrl(url, https) {
    // remove .git suffix
    url = url.replace(/\.git$/, "");
    // normalize git@host urls
    if (url.startsWith("git@")) {
        url = url.replace(/^git@([^:]+):(.*)$/, (https ? "https" : "http") + "://$1/$2");
    }
    // remove trailing slashes
    url = url.replace(/\/+$/, "");
    return new URL(url);
}
function getRemoteUrl(https = true) {
    try {
        const file = (0, mod_js_1.join)(dntShim.Deno.cwd(), ".git", "config");
        const content = dntShim.Deno.readTextFileSync(file);
        const data = (0, mod_js_3.parse)(content);
        const origin = data['remote "origin"'];
        if (!origin?.url) {
            return;
        }
        return normalizeUrl(origin.url, https).href;
    }
    catch (err) {
        console.error(red(err.message));
        // Ignore
    }
}
function showHelp() {
    console.log(`keep-a-changelog

Usage: keep-a-changelog [options]

Options:
  --file, -f                Changelog file (default: CHANGELOG.md)
  --format                  Output format (default: compact)
  --bullet-style            Bullet point style (default: -)
  --url                     Repository URL

  --init                    Initialize a new changelog file
  --latest-release          Print the latest release version
  --latest-release-full     Print the latest release

  --release                 Set the date of the specified release
  --combine                 Combine changes from releases with the same version
  --create                  Create a new release. Optionally accepts a version number or 'patch', 'minor' or 'major'

  --no-v-prefix             Do not add a "v" prefix to the version
  --no-sort-releases        Do not sort releases
  --head                    Set the HEAD link
  --quiet                   Do not print errors
  --help, -h                Show this help message
`);
}
